packages/core/src/api/blockManipulation/commands/updateBlock/updateBlock.ts

```diff
@@ -36,135 +39,202 @@ export const updateBlockCommand =
+      | undefined;
+    blockContent: { node: PMNode; beforePos: number; afterPos: number };
+  }
+) {
+  let content: PMNode[] | "keep" = "keep";
+
+  // Has there been any custom content provided?
+  if (block.content) {
+    if (typeof block.content === "string") {
+      // Adds a single text node with no marks to the content.
+      content = inlineContentToNodes(
+        [block.content],
+        state.schema,
+        editor.schema.styleSchema
+      );
+    } else if (Array.isArray(block.content)) {
+      // Adds a text node with the provided styles converted into marks to the content,
+      // for each InlineContent object.
+      content = inlineContentToNodes(
+        block.content,
+        state.schema,
+        editor.schema.styleSchema
+      );
+    } else if (block.content.type === "tableContent") {
+      content = tableContentToNodes(
+        block.content,
+        state.schema,
+        editor.schema.styleSchema
+      );
+    } else {
+      throw new UnreachableCaseError(block.content.type);
+    }
+  } else {
+    // no custom content has been provided, use existing content IF possible
+    // Since some block types contain inline content and others don't,
+    // we either need to call setNodeMarkup to just update type &
+    // attributes, or replaceWith to replace the whole blockContent.
+    if (oldNodeType.spec.content === "") {
+      // keep old content, because it's empty anyway and should be compatible with
+      // any newContentType
+    } else if (newNodeType.spec.content !== oldNodeType.spec.content) {
+      // the content type changed, replace the previous content
+      content = [];
+    } else {
+      // keep old content, because the content type is the same and should be compatible
+    }
+  }
+
+  // Now, changes the blockContent node type and adds the provided props
+  // as attributes. Also preserves all existing attributes that are
+  // compatible with the new type.
+  //
+  // Use either setNodeMarkup or replaceWith depending on whether the
+  // content is being replaced or not.
+  if (content === "keep") {
+    // use setNodeMarkup to only update the type and attributes
+    state.tr.setNodeMarkup(
+      blockInfo.blockContent.beforePos,
+      block.type === undefined ? undefined : state.schema.nodes[block.type],
+      {
+        ...blockInfo.blockContent.node.attrs,
+        ...block.props,
+      }
+    );
+  } else {
+    // use replaceWith to replace the content and the block itself
+    // also  reset the selection since replacing the block content
+    // sets it to the next block.
+    state.tr.replaceWith(
+      blockInfo.blockContent.beforePos,
+      blockInfo.blockContent.afterPos,
+      newNodeType.create(
+        {
+          ...blockInfo.blockContent.node.attrs,
+          ...block.props,
+        },
+        content
+      )
+    );
+    // TODO: This seems off - the selection is not necessarily in the block
```

Comment:
@matthewlipski remove this comment + maybe quick sanity check if there were merge issues